content/en/blog/_posts/2024-12-12-scheduler-queueinghint/index.md

```diff
@@ -0,0 +1,124 @@
+
+If the scheduling is successful, the second phase, called the _binding cycle_, binds the Pod with
+the Node by communicating the decision to the API server.
+But, if it turns out that the Pod cannot go to anywhere during the scheduling cycle,
+the binding cycle isn't executed; instead the Pod is moved back to the scheduling queue.
+Although there are some exceptions, unscheduled Pods enter the _unschedulable pod pool_.
+
+Pods in Unschedulable Pod Pool are moved to ActiveQ/BackoffQ 
+only when Scheduling Queue identifies changes in the cluster that might be schedulable if we retry the scheduling.
+
+That is a crucial step because scheduling cycle is performed for Pods one by one -
+if we didn't have Unschedulable Pod Pool and kept retrying the scheduling of any Pods, 
+multiple scheduling cycles would be wasted for Pods that have no chance to be scheduled.
+
+Then, how do they decide when to move a Pod back into the ActiveQ? How do they notice that Pods might be schedulable now?
+Here QueueingHints come into play.
+
+## QueueingHint
+
+QueueingHint is callback function per plugin to notice an object addition/update/deletion in the cluster (we call them cluster events)
+that may make Pods schedulable.
+
+Let's say the Pod `pod-a` has a required Pod affinity, and got rejected in scheduling cycle by the `PodAffinity` plugin
+because no Node has any Pod matching the Pod affinity specification for `pod-a`.
+
+![pod-a got rejected by PodAffinity](./queueinghint1.svg)
+
+When an unscheduled Pod is put into the unschedulable pod pool, the scheduling queue
+records which plugins caused the scheduling failure of the Pod.
+In this example, scheduling queue notes that `pod-a` was rejected by `PodAffinity`.
+
+`pod-a` will never be schedulable until the PodAffinity failure is resolved somehow.
+The scheduling queue uses the queueing hints from plugins that rejected the Pod, which is `PodAffinity` in the example.
+
+A QueueingHint subscribes to a particular kind of cluster event and make a decision whether an incoming event could make the Pod schedulable.
+Thinking about when PodAffinity failure could be resolved,
+one possible scenario is that an existing Pod gets a new label which matches with `pod-a`'s PodAffinity.
+
+The `PodAffinity` plugin's `QueueingHint` callback checks on all Pod updates happening in the cluster, 
+and when it catches such update, the scheduling queue moves `pod-a` to either ActiveQ or BackoffQ.
+
+![pod-a is moved by PodAffinity QueueingHint](./queueinghint2.svg)
+
+We actually already had a similar functionality (called `preCheck`) inside the scheduling queue,
+which filters out cluster events based on Kubernetes core scheduling constraints - 
+for example, filtering out node related events when nodes aren't ready.
+
+But, it's not ideal because this hard-coded `preCheck` refers to in-tree plugins logic,
+and it causes issues for custom plugins (for example: [#110175](https://github.com/kubernetes/kubernetes/issues/110175)).
+
```

Comment:
```suggestion
## Improvements to retrying Pod scheduling with QueuingHint

Unschedulable Pods only move back into the ActiveQ or BackoffQ components of the scheduler
queue if changes in the cluster might allow the scheduler to place those Pods on nodes. Prior to v1.32, 
the scheduling queue identified these changes in the cluster by using a feature called `preCheck`. This
built-in functionality filters out cluster events based on Kubernetes core scheduling constraints. For example,
`preCheck` could filter out node-related events when the node status is `NotReady`. 

However, `preCheck` relies on the logic of in-tree plugins and caused some issues with custom plugins,
like in issue [#110175](https://github.com/kubernetes/kubernetes/issues/110175). Additionally, the broad
scope of event filtering could lead to scheduling retries for no reason. For example, the
scheduling queue might retry scheduling after a node update event in case the update created more
resources for the previously unschedulable Pod. However, this would result in a retry attempt for every
node update, even if those updates didn't actually change resource capacity on the nodes. 

QueueingHint is a per-plugin callback function to notice an object creation, update, or deletion in the cluster,
called _cluster events_, that might make unschedulable Pods schedulable. A QueueingHint subscribes to a 
particular kind of cluster event and make a decision about whether an incoming event could make the Pod
schedulable. The scheduling queue uses the queueing hints from plugins that rejected the Pod to decide
whether to retry scheduling for that Pod. 

For example, consider a Pod named `pod-a` that has a required Pod affinity. `pod-a` was rejected in
the scheduling cycle by the `PodAffinity` plugin because no node had an existing Pod that matched
the Pod affinity specification for `pod-a`.

![pod-a got rejected by PodAffinity](./queueinghint1.svg)

`pod-a` moves into the Unschedulable Pod Pool. The scheduling queue records which plugin caused
the scheduling failure for the Pod. For `pod-a`, the scheduling queue records that the `PodAffinity`
plugin rejected the Pod.

`pod-a` will never be schedulable until the PodAffinity failure is resolved. The `PodAffinity` plugin's
`QueuingHint` callback function checks every Pod update that occurs in the cluster. If, for example,
a Pod gets a label update that matches the Pod affinity requirement of `pod-a`, the `PodAffinity`
plugin's `QueuingHint` prompts the scheduling queue to move `pod-a` back into the ActiveQ or
the BackoffQ component.

![pod-a is moved by PodAffinity QueueingHint](./queueinghint2.svg)
```

Reply:
Thanks, but some explanations are not quite correct and I had to modify.